### PR TITLE
Add telemetry to bundler

### PR DESCRIPTION
**Pull request by garryod**
Add telemetry (logs + metrics + traces) to bundler using [`tracing`](https://docs.rs/tracing/latest/tracing/) with export in OpenTelemetry format via [`opentelemetry-otlp`](https://docs.rs/opentelemetry-otlp/latest/opentelemetry_otlp/).

Collecting & visualising results with:
- Jaeger for traces
- Prometheus for metrics

### CODE REVIEW

**Comment by garryod** [comment]
> What does Jaeger offer over something like Grafana that's used elsewhere at Diamond? Are we going to end up with an individual monitoring system for each service or is the aim that they're all going to be combined into a central system at some point?

Jaeger faciliates distributed tracing, as opposed to metrics visualisation. That is to say it provides a view of 'events' each of which can be drilled down into giving a waterfall of spans which were involved in this event and logs which were captured in those spans, even across service boundaries.

To really make distributed tracing powerful there needs to be a single instance which can correlate spans from desperate services, so I would envisage having one instance (cluster for HA) which all services would push their tracing data to.

File: .devcontainer/docker-compose.yml [hunk 1]
```diff
@@ -12,6 +12,7 @@ services:
       DATABASE_URL: mysql://root:rootpassword@ispyb/ispyb_build
       BUNDLER_DATABASE_URL: mysql://root:rootpassword@ispyb/ispyb_build
       BUNDLER_LOG_LEVEL: DEBUG
+      BUNDLER_OTEL_COLLECTOR_URL: http://collector:4317
```
[the remark this answers is not in the document]
Comment (garryod) [reply .devcontainer/docker-compose.yml:15]: `otelcol` & `jaeger` both support env var substitution in their configs, so I could put them in the `docker-compose`. Unfortunately `prometheus` doesn't

**Comment by tpoliaw** [comment]
I thought grafana offered tracing visualisation as well but looking at it, [it requires an external data source (such as jaeger) to work](https://grafana.com/docs/grafana/latest/panels-visualizations/visualizations/traces/). I have no preferences one way or the other, I'm just not sure of what each part is doing. Where does prometheus fit in?

**Review by DiamondJoseph — APPROVED** [review]
If Peter is happy with the Rust, the devcontainer changes look sensible; presumably we have an equivalent standard logging deployment which handles the jaeger?

**Comment by garryod** [comment]
> If Peter is happy with the Rust, the devcontainer changes look sensible; presumably we have an equivalent standard logging deployment which handles the jaeger?

Nothing in the deployment yet, and no recommended good practice for how to do this, kind of using this project (and XChem) as the test bed for OTEL & Jaeger with the view to spin up some Diamond wide infra to handle traces & metrics

**Pull request merged**